### PR TITLE
Migrate to cirrus-ci

### PR DESCRIPTION
**Pull request by bjlittle**
### Description
This PR migrates `iris` from using `travis-ci` to `cirrus-ci`.

It also introduces the use of [nox](https://nox.thea.codes/en/stable/index.html) as a consistent entry point for test automation, both for CI and interactively as a developer. The use of `nox` also facilities clear namespace separation between CI infra-structure and testing capability. 

The use of `cirrus-ci` also allows the use of caching for:
- the CI conda environment
- the nox test environments
- `iris-test-data`
- `cartopy` shapefiles

Cache generation, uploading and downloading is automated. In addition, controls have been added to manually invalidate a cache, and caches will be refreshed after a configurable time period (to ensure that the contents of the cache is not stale, relative to outside sources)

---
[Consult Iris pull request check list]( https://scitools-iris.readthedocs.io/en/latest/developers_guide/contributing_pull_request_checklist.html)

### CODE REVIEW

File: lib/iris/tests/test_basic_maths.py
Comment (bjlittle) [inline lib/iris/tests/test_basic_maths.py:558]: Yeah, can't explain this failure when run on CI - it's not repeatable locally.

**Comment by bjlittle** [comment]
This PR also includes automated `flake8` linting. So we could turn off `stickler-ci`

**Comment by bjlittle** [comment]
@tkknight The documentation link check is consistently failing for particular URLs... We could update the URL skip list?

File: lib/iris/tests/test_basic_maths.py
Comment (bjlittle) [reply lib/iris/tests/test_basic_maths.py]: Okay, discovered that this failure was the result of dtype quantization, given the test data being used in the maths operation.

In summary, the `(73, 96)` of `float32` data resulted in `4` values out of `7,008` being inaccurate with a tolerance of approximately `e-07`.

Simply increasing the `dtype` resolves this, hence a different CML data checksum.

**Comment by bjlittle** [comment]
As a strategy, I think that this PR should target the `3.0.x` branch instead of `master`.

Once merged into `3.0.x`, we can then merge back the `3.0.x` changes to `master`.

**Comment by trexfeathers** [comment]
> This PR also includes automated `flake8` linting. So we could turn off `stickler-ci`.
> 
> Does anyone have an opinion on that?

The fewer tools the better IMO

**Comment by bjlittle** [comment]
Note that, I've not replicated the `travis-ci` step to [publish the docs to github pages](https://github.com/SciTools/iris/blob/master/.travis.yml#L161-L169).

I'm raising the question as to whether this is necessary now that we have `readthedocs-ci` enabled.

@SciTools/iris-devs Isn't this now redundant?

File: setup.cfg
Comment (trexfeathers) [inline setup.cfg:1]: Was nothing using this content, with everything using [.flake8](https://github.com/bjlittle/iris/blob/migrate-to-cirrus-ci/.flake8) instead?

As I understand it, this PR has so far only _added_ to the tools running `flake8` (Stickler is still being used at the time of writing).

Comment (bjlittle) [reply setup.cfg:1]: Yup, this wasn't needed. The `flake8` configuration is in the `.flake8` file instead

Comment (trexfeathers) [reply setup.cfg:1]: It appears I wrote this at the same time as @bjlittle removed Stickler (5dfd8a160f899bd0bfc2e7379fd3fba12db8bcc4). Does that explain this removal?

Comment (trexfeathers) [reply setup.cfg:1]: Thanks

**Comment by tkknight** [comment]
> @tkknight The documentation link check is consistently failing for particular URLs... We could update the URL skip list?

Sure.  Looks like only 1 link needs ignoring:  http://www.personal.psu.edu/cab38/ColorBrewer/ColorBrewer_updates.html

The link loads fine in a web browser.

**Comment by tkknight** [comment]
> Note that, I've not replicated the `travis-ci` step to [publish the docs to github pages](https://github.com/SciTools/iris/blob/master/.travis.yml#L161-L169).
> 
> I'm raising the question as to whether this is necessary now that we have `readthedocs-ci` enabled.
> 
> @SciTools/iris-devs Isn't this now redundant?

I think so too.  The content on the github pages can just stay as it is showing the docs for Iris 2.4 and earlier.

File: .cirrus.yml
Comment (tkknight) [inline .cirrus.yml:60]: Is this ampersand correct?

File: .cirrus.yml
Comment (tkknight) [inline .cirrus.yml:106]: what does the ``<< : *LINUX_TASK_TEMPLATE`` mean? (just curious)

File: .cirrus.yml
Comment (tkknight) [inline .cirrus.yml:109]: Note to self - there are a few occurrences of the ``docs`` directory being used for the cirrus-ci setup.

File: .cirrus.yml
Comment (bjlittle) [reply .cirrus.yml:60]: Yup. See https://cirrus-ci.org/guide/tips-and-tricks/#sharing-configuration-between-tasks

Essentially, it's a mechanism to use YAML aliases in order to share common YAML in several places. It's pretty cool.

In summary, there are two parts to this. The `&` part and the `*` part below... it's kinda like jinja templating, for template definition (`&`) and then template use (`*`).

Does that help?

File: .cirrus.yml
Comment (bjlittle) [reply .cirrus.yml:106]: @tkknight See above :+1:

**Comment by tkknight** [comment]
More linkchecks are failing, snippet below:

``(line  283) broken    https://github.com/rcomer - 429 Client Error: too many requests for url: https://github.com/rcomer``
``(line  315) broken    https://github.com/lbdreyer - 429 Client Error: too many requests for url: https://github.com/lbdreyer``
``(line  327) broken    https://github.com/OSGeo/PROJ/pull/1292 - 429 Client Error: too many requests for url: https://github.com/OSGeo/PROJ/pull/1292``

Looks like cirrus is hitting github a lot (likely not just the iris project) and this the checks are broken.  I suggest ignoring github domain fully, ie:

``https://github.com``

**Review by trexfeathers — CHANGES_REQUESTED** [review]
Removing Travis and making nox the entry point for CI and local testing requires updates to several pages in the [developer's guide](https://github.com/SciTools/iris/tree/master/docs/iris/src/developers_guide), and possibly other parts of the documentation.

**Comment by tkknight** [comment]
> Removing Travis and making nox the entry point for CI and local testing requires updates to several pages in the [developer's guide](https://github.com/SciTools/iris/tree/master/docs/iris/src/developers_guide), and possibly other parts of the documentation.

Yup, have already suggested to @bjlittle that this be a separate PR to we can get the ci up and running asap.

**Comment by bjlittle** [comment]
I'm going to retarget `3.0.x` instead of `master` now that the tests are passing :tada:

**Comment by rcomer** [comment]
If `nox` is to be run for local testing, and includes both `flake8` and `black`, do we still need `flake8` and `black` in the pre-commit hooks?

**Comment by bjlittle** [comment]
> If `nox` is to be run for local testing, and includes both `flake8` and `black`, do we still need `flake8` and `black` in the pre-commit hooks?

@rcomer Good question. I think the answer is yes. Some people use the pre-commit hooks, some don't - personally I do. The end result is pretty much the same, but having the pre-commit hooks gives me confidence that what I've committed is fit for purpose without having to explicitly check it myself (yup, I'm fundamentally lazy :smile:)

You could choose not to have `flake8` and `black` as part of the pre-commit hook, and `nox` would catch it as part of the CI, but that for me is a slower turn around than discovering lint and style issues immediately and locally.

You could run `nox` locally e.g., `nox -s flake8 black`, and that would do the job, but personally I'm more than happy to delegate that the pre-commit hook as I commit (liek I said, I'm lazy).

This all seems like unnecessary duplication, but I think that people work in different ways, so empowering the developer with options (rather than railroad them with a single workflow) seems quite attractive and flexible to me.

**Comment by bjlittle** [comment]
I'm done testing and tweaking... I think this PR is good to go, pending review.

I've added a couple of associated `whatsnew` entries. You can view the rendered results [here](https://bjlittle-iris.readthedocs.io/en/latest/whatsnew/3.0.html#internal).

I'll follow-up this PR with the necessary documentation changes. And when that's merged, we can merge back the `v3.0.x` branch to `master` in order to enable `cirrus-ci` goodness there :wink:

**Comment by tkknight** [comment]
Looks good @bjlittle.  Thanks for all the effort on this.

**Pull request merged**
